src/prognostic_equations/edmfx_entr_detr.jl

```diff
@@ -0,0 +1,284 @@
+    ::NoDetrainment,
+) where {FT}
+    return FT(0)
+end
+
+function pi_groups_detr(
+    params,
+    ᶜz::FT,
+    z_sfc::FT,
+    ᶜp::FT,
+    ᶜρ::FT,
+    buoy_flux_surface::FT,
+    ᶜaʲ::FT,
+    ᶜwʲ::FT,
+    ᶜRHʲ::FT,
+    ᶜbuoyʲ::FT,
+    ᶜw⁰::FT,
+    ᶜRH⁰::FT,
+    ᶜbuoy⁰::FT,
+    dt::FT,
+    ::PiGroupsDetrainment,
+) where {FT}
+
+    if ᶜaʲ <= FT(0)
+        return FT(0)
+    else
+        g = CAP.grav(params)
+        turbconv_params = CAP.turbconv_params(params)
+        ᶜaʲ_max = TCP.max_area(turbconv_params)
+        max_area_limiter = 0.1 * exp(-10 * (ᶜaʲ_max - ᶜaʲ))
+
+        # pressure scale height (height where pressure drops by 1/e)
+        ref_H = ᶜp / (ᶜρ * g)
+        # convective velocity
+        w_star = get_wstar(buoy_flux_surface)
+
+        # non-dimensional pi-groups
+        # TODO - using Π₁ blows things up
+        Π₁ =
+            (ᶜz - z_sfc) * (ᶜbuoyʲ - ᶜbuoy⁰) /
+            ((ᶜwʲ - ᶜw⁰)^2 + w_star^2 + eps(FT))
+        Π₃ = sqrt(ᶜaʲ)
+        Π₄ = ᶜRHʲ - ᶜRH⁰
+        Π₆ = (ᶜz - z_sfc) / ref_H
+        detr = max(
+            0,
+            min(
+                abs(ᶜwʲ) * (
+                    3.535208 + 0.598496 * Π₁ + 1.583348 * Π₃ + 0.046275 * Π₄ - 0.344836 * Π₆ + max_area_limiter
+                ),
+                1 / dt,
+            ),
+        )
+        return detr
+    end
+end
+
+function pi_groups_detr(
+    params,
+    ᶜz::FT,
+    z_sfc::FT,
+    ᶜp::FT,
+    ᶜρ::FT,
+    buoy_flux_surface::FT,
+    ᶜaʲ::FT,
+    ᶜwʲ::FT,
+    ᶜRHʲ::FT,
+    ᶜbuoyʲ::FT,
+    ᶜw⁰::FT,
+    ᶜRH⁰::FT,
+    ᶜbuoy⁰::FT,
+    dt::FT,
+    ::ConstantCoefficientDetrainment,
+) where {FT}
+
+    if ᶜaʲ <= FT(0)
+        return FT(0)
+    else
+        detr_coeff = CAP.detr_coeff(params)
+        detr = max(0, min(detr_coeff * abs(ᶜwʲ), 1 / dt))
+        return detr
+    end
+end
+
+edmfx_entr_detr_tendency!(Yₜ, Y, p, t, colidx, turbconv_model) = nothing
+function edmfx_entr_detr_tendency!(Yₜ, Y, p, t, colidx, turbconv_model::EDMFX)
+
+    n = n_mass_flux_subdomains(turbconv_model)
+    (; ᶜspecificʲs, ᶜh_totʲs, ᶜentrʲs, ᶜdetrʲs) = p
+    (; ᶜu⁰, ᶜspecific⁰, ᶜh_tot⁰) = p
+
+    for j in 1:n
+
+        @. Yₜ.c.sgsʲs.:($$j).ρa[colidx] +=
+            Y.c.sgsʲs.:($$j).ρa[colidx] *
+            (ᶜentrʲs.:($$j)[colidx] - ᶜdetrʲs.:($$j)[colidx])
+
+        @. Yₜ.c.sgsʲs.:($$j).ρae_tot[colidx] +=
+            Y.c.sgsʲs.:($$j).ρa[colidx] * (
+                ᶜentrʲs.:($$j)[colidx] * ᶜh_tot⁰[colidx] -
+                ᶜdetrʲs.:($$j)[colidx] * ᶜh_totʲs.:($$j)[colidx]
+            )
+
+        @. Yₜ.c.sgsʲs.:($$j).ρaq_tot[colidx] +=
+            Y.c.sgsʲs.:($$j).ρa[colidx] * (
+                ᶜentrʲs.:($$j)[colidx] * ᶜspecific⁰.q_tot[colidx] -
+                ᶜdetrʲs.:($$j)[colidx] * ᶜspecificʲs.:($$j).q_tot[colidx]
+            )
+
+        @. Yₜ.f.sgsʲs.:($$j).u₃[colidx] +=
+            ᶠinterp(ᶜentrʲs.:($$j)[colidx] * C3(ᶜu⁰[colidx])) -
+            ᶠinterp(ᶜentrʲs.:($$j)[colidx]) * Y.f.sgsʲs.:($$j).u₃[colidx]
```

Comment:
Would it make sense to put the entr/detr TKE part also here? So that we have it all in one place?

Reply:
We decided to think about where to put it when adding the TKE equation for prognostic EDMF.